azure_functions_worker/functions.py

```diff
@@ -243,21 +230,70 @@ def add_function(self, function_id: str,
                     f'annotation {return_pytype!r}')
 
             if return_pytype is (str, bytes):
-                return_binding_name = 'generic'
+                binding_name = 'generic'
 
-            if not bindings.check_output_type_annotation(
-                    return_binding_name, return_pytype):
+            if not bindings_utils.check_output_type_annotation(
+                    binding_name, return_pytype):
                 raise FunctionLoadError(
                     func_name,
                     f'Python return annotation "{return_pytype.__name__}" '
-                    f'does not match binding type "{return_binding_name}"')
+                    f'does not match binding type "{binding_name}"')
 
         if has_implicit_return and 'return' in annotations:
             return_pytype = annotations.get('return')
 
         return_type = None
         if has_explicit_return or has_implicit_return:
-            return_type = ParamTypeInfo(return_binding_name, return_pytype)
+            return_type = ParamTypeInfo(binding_name, return_pytype)
+
+        return return_type
+
+    def add_function(self, function_id: str,
+                     func: typing.Callable,
+                     metadata: protos.RpcFunctionMetadata):
+        func_name = metadata.name
+        sig = inspect.signature(func)
+        params = dict(sig.parameters)
+        annotations = typing.get_type_hints(func)
+        return_binding_name: typing.Optional[str] = None
+        has_explicit_return = False
+        has_implicit_return = False
+
+        bound_params = {}
+        for binding_name, binding_info in func.metadata.bindings.items():
```

Comment:
line 265 to 280 duplicate with add_indexed_function, recommend extract to a function